### ISSUE TITLE
Better default director:except for Python

### ISSUE THREAD

**Issue by letmaik** [swig/swig#1117]
The [current default](https://github.com/swig/swig/blob/4f4147a779bdf2c2a323b862ba7c9941d1bf7729/Source/Modules/python.cxx#L5647-L5649) for `%feature("director:except")` is:

```c++
Append(w->code, "  if (error) {\n");
Printf(w->code, "    Swig::DirectorMethodException::raise(\"Error detected when calling '%s.%s'\");\n", classname, pyname);
Append(w->code, "  }\n");
```

Which means that by default, any Python-level exceptions occurring within director class methods will not have any error message from the Python exception propagated which is a massive pain for users.

My current new default is the following:

```c++
%feature("director:except") {
  if ($error != NULL) {
    PyObject *exc, *val, *tb;
    PyErr_Fetch(&exc, &val, &tb);
    PyErr_NormalizeException(&exc, &val, &tb);
    std::string err_msg("In method '$symname': ");

    PyObject* exc_str = PyObject_GetAttrString(exc, "__name__");
    err_msg += PyUnicode_AsUTF8(exc_str);
    Py_XDECREF(exc_str);
    
    if (val != NULL)
    {
      PyObject* val_str = PyObject_Str(val);
      err_msg += ": ";
      err_msg += PyUnicode_AsUTF8(val_str);
      Py_XDECREF(val_str);
    }

    Py_XDECREF(exc);
    Py_XDECREF(val);
    Py_XDECREF(tb);
    
    Swig::DirectorMethodException::raise(err_msg.c_str());
  }
}
```
This leads to Python exception messages like:
> SWIG director method error. In method 'foo': NameError: name 'x' is not defined

Something along those lines should be the default for Python I think. This is specifically for Python 3, so it has to be adapted to be Python 2 compatible.

Ideally, the error message should not mention "SWIG director method error." and be just the plain Python exception "NameError: name 'x' is not defined" with proper stacktrace, but maybe this goal is a lot harder to achieve.

**Comment by vadz** [comment]
This is (supposed to be) fixed by https://github.com/swig/swig/pull/1096

I didn't have time/motivation to add a unit test for this yet, if you can help with this to allow that PR to be merged, it would be appreciated.

**Comment by letmaik** [comment]
I don't have the resources currently to understand how to write swig tests, but I was thinking about something else. Since it is a Python-level exception, wouldn't it make sense to keep the Python stack trace alive? In fact, why not just throw the actual Python exception and skip all the DirectorMethodException business? Otherwise it's just too confusing for Python users as they don't see the line where the error occurred.

**Comment by letmaik** [comment]
I figured out a better way to handle this situation without having to extract Python exception details and wrap them again.

First: This is a no-problem if the user hasn't defined %exception. In this case, a director method exception will be caught in generated code like that:
```c
  director = SWIG_DIRECTOR_CAST(arg1);
  upcall = (director && (director->swig_get_self()==obj0));
  try {
    if (upcall) {
      (arg1)->MyClass::foo(arg2,arg3);
    } else {
      (arg1)->foo(arg2,arg3);
    }
  } catch (Swig::DirectorException&) {
    SWIG_fail;
  }
```
This leads to the control flow immediately being passed back to Python which then displays the regular Python exception, since it hasn't been modified or cleared within the SWIG code. Great!

However, often people have a global exception handler like:
```swig
/*%exception {
  try {
    $action
  } catch (const std::exception& e) {
    SWIG_exception(SWIG_RuntimeError, e.what());
  }
}
```

In this case, the generated code looks like:
```c
  director = SWIG_DIRECTOR_CAST(arg1);
  upcall = (director && (director->swig_get_self()==obj0));
  try {
    {
      try {
        if (upcall) {
          (arg1)->MyClass::foo(arg2,arg3);
        } else {
          (arg1)->foo(arg2,arg3);
        }
      } catch (const std::exception& e) {
        SWIG_exception(SWIG_RuntimeError, e.what());
      }
    }
  } catch (Swig::DirectorException&) {
    SWIG_fail;
  }
```
And this is bad! SWIG_exception(..) overrides the last Python exception with a new one and there is no way to recover from that in a nice way.
The work-around I posted initially using `%feature("director:except")` tries to embed as much information as possible in this newly thrown exception, but that misses the point and only fights the symptoms, compared to fixing the root cause.

The current solution to have both a catch-all error handler and pass through Python-level director method exceptions is to extend `%exception` like that:
```swig
%exception {
  try {
    $action
  } catch (Swig::DirectorMethodException&) {
    SWIG_fail;
  } catch (const std::exception& e) {
    SWIG_exception(SWIG_RuntimeError, e.what());
  }
}
```

I have a feeling this should work out of the box even with custom global exception handlers but I can't come up with a good solution for it. A temporary solution may be to add a big warning in the director section of the SWIG Python documentation about exactly this and that if `%exception` is used, then it must have that special first catch.

**Comment by vadz** [comment]
> The work-around I posted initially using `%feature("director:except")` tries to embed as much information as possible in this newly thrown exception, but that misses the point and only fights the symptoms, compared to fixing the root cause.

I'm really not sure why do you say this. AFAICS the root cause here is that we don't return the Python error that occurs inside `PyObject_CallFunction()` (or `PyObject_CallMethod()`) that we call. And we must have the code you added (and that I did as well in #1096), i.e. call `PyErr_Fetch()` somewhere. I don't see how can it work without it.

**Comment by letmaik** [comment]
My goal was to not wrap the Python exception in a new SWIG exception, but rather throw the original Python exception as-is back to the user. By doing `PyErr_Fetch()` you clear the exception and would have to call `PyErr_Restore()` later on to recover the original exception. Preserving the Python stacktrace is I think only possible by not fetching and wrapping the Python error.

**Comment by vadz** [comment]
Wait, but how could this be at all possible? Python code calls C++ code which calls Python code, which generates the exception -- there is no way to return this exception directly to Python, is there?

**Comment by letmaik** [comment]
It just seems to work! :) And the stacktrace looks alright, it includes the stack frames from before and after the C++ part. It kind of makes sense, if no one is overriding the current Python error then the right thing to do is just return from C++ back to Python and let Python handle the error which is set as active error via the PyErr_ error indicators. If you handle the error in C++ then this is equivalent to try-except and rethrowing it, which is absolutely fine but not what most people want I think.

**Comment by vadz** [comment]
Huh, I guess I see how it's possible, after all (the interpreter global error variable just remains unchanged through all that...), but I'm not sure if it's really wise to rely on this. OTOH I agree that it's very nice to have the correct stack.

**Comment by letmaik** [comment]
Not being able to see the correct file/line of the error was brought up as number one issue by researchers in one of the projects I work. Especially since Python is such a dynamic language where you very often make errors that only surface at runtime, debugging that without stacktrace is nearly impossible, especially when the director method is longer than a few lines.

**Comment by wsfulton** [comment]
I agree that we do need a better default and the default should be just to through the original exception. The behaviour should be fully customisable too, for example it should be possible to not have any C++ exceptions as requested in #315.

I believe the Java implementation satisifies all these requirements. I've just committed some changes to fully document it all and also add in a missing gap of showing the original stack trace by default. Please look at it at https://htmlpreview.github.io/?https://github.com/swig/swig/blob/master/Doc/Manual/Java.html#Java_exceptions_from_directors. Note the `%catches(Swig::DirectorException)` for working with a global `%exception`. Also note that no C++ exception catch handlers are generated by default (as per non-director wrappers), a user *has* to add these in via `%catches`or `%exception`. The full  customisations are quite complex to grasp, I think, but it is at least possible. Besides, full customisation will rarely be needed as the default (with the simple `%catches`) is very good and should satisfy the vast majority of users.

I suggest all the different languages are modified to work in a similar way. The only thing that needs some target language specifics is how to clear the target language exception, store it, then rethrow it. @letmaik's approach is to leave the exception pending, which I think is not ideal, as a user may want to use the C Python API before returning to the interpreter (which doesn't feel right with a pending exception). Certainly in Java, this approach doesn't work.

**Comment by degasus** [comment]
I have been hit by this issue as well. I can confirm that rethrowing the exception works fine in the `python -> C++ -> python` case. However if the `C++ -> python` call happens on another thread, forwarding the exception is not that trivial (at least I was not able to manage it). letmaik's code works fine, however I would like to suggest to replace the what() method with this code.

**Comment by ojwb** [comment]
Reading the above it strikes me that this would work better if `Swig::DirectorException` did **not** derive from `std::exception` (because then it wouldn't be caught by a user's `%exception` block which catches and handles `std::exception` but would reach the outer catch that's specifically for `Swig::DirectorException`).

Clearly that's would not be entirely compatible with how things are currently, but it is perhaps something we could consider for SWIG 4.1.

**Comment by degasus** [comment]
@ojwb Please be aware that the director may not be called from an outer python function. So there might be no outer catch block.

**Comment by ojwb** [comment]
If that's the case, then currently won't the `Swig::DirectorException` only get caught if the user happens to have provided a `%exception` block which catches `std::exception`?  (Or which catches `Swig::DirectorException` explicitly, but in that case my suggested base class change doesn't matter anyway.)

**Comment by degasus** [comment]
In my case, the %exception block has no effect. I have a C++ thread, which calls a virtual method, provided by a python method with swig directors. Within C++, the general `catch std::exception&` clause is fine right now. Catching the `DirectorException` might be better, however I would like to avoid sharing this custom exception class with other shared libraries.
